baseservice/iodevices/input_devices.py

```diff
@@ -0,0 +1,399 @@
+from abc import ABCMeta, abstractmethod
+from threading import Event
+from time import time, sleep
+from typing import Optional, Tuple, List, Dict, Union
+
+from baseservice.iodevices.common import Message, DeviceHeaders
+from baseservice.utils import KwargsException, StatefulListIterator
+
+
+class InputDeviceException(KwargsException):
+    """
+    a base exception class for all input device related exceptions
+    """
+    pass
+
+
+class InputTransaction(metaclass=ABCMeta):
+    """
+    this is the base class for input transaction objects.
+
+    after reading a message from the device (using 'with_transaction=True') the reader should commit/rollback
+    the transaction, to signal the device if the message is done processing or not
+    """
+
+    def __init__(self, device: 'InputDevice'):
```

Comment:
Instead of using types as strings, the newer way is to import annotations which will allow you to do the same without the string

Reply:
i don't understand what you mean. you're welcome to show me an example

Reply:
```suggestion
from __future__ import annotations

    def __init__(self, device: InputDevice):
```

Reply:
what is it good for? what's wrong with the strings? i prefer not to add extra 'things' if it doesn't help ALOT

Reply:
Its just the newer way of approaching this. Its a feature that will become a standard in a future python release.